c++/src/kj/memory.h

```diff
@@ -594,6 +607,13 @@ Own<Decay<T>> heap(T&& orig) {
   return Own<T2>(new T2(kj::fwd<T>(orig)), _::HeapDisposer<T2>::instance);
 }
 
+template <typename T, void(*F)(T*)>
```

Comment:
Hmm, I guess using these template parameters implies that callers will always have to redundantly specify T, like:

    auto own = rawPtr<MyType, &freeMyType>(ptr);

But specifying `MyType` here is redundant, since it should be deduced from `ptr`. Ideally the syntax would be more like:

    auto own = rawPtr<&freeMyType>(ptr);

But I'm not sure how to declare a template that would accept this. Is it possible? I know that the following would be possible, but it's weird:

    // too weird
    auto own = rawPtr(ptr).withDisposer<&freeMyType>();

(I've run into this difficulty a few times in the past, and I swear I came across a solution once, but I've forgotten what it was...)

Reply:
> auto own = rawPtr<&freeMyType>(ptr);

Looks like we can achieve this (although `&` not necessary due to decay to function pointer) using:

```
template <auto F, typename T>
Own<T> rawPtr(T* ptr) {
...
```

and keeping the implementation the same, which lets the `CustomDisposer` enforce the constraints on F.

Unfortunately, that's C++17 only and it looks like capnproto is still on C++14?

Reply:
Oh nice, why didn't I know about that?

I'm happy to increase the requirement to C++17, or maybe even C++20 (for coroutines!), but probably not in this PR.

For now, you can wrap this in:

`#if __cplusplus > 201402L`

(We have some other features that are C++17-only like this, such as `kj::arr()` to construct a fixed-size array.)

Reply:
Done.